loadtest.go

```diff
@@ -89,11 +91,114 @@ func (l *loadTest) runUser() {
 	l.log.Debug("user stopped")
 }
 
+func registerPushDevice(config *config.Config, deviceID, outputChannel string, rest *ably.REST, log log15.Logger) error {
+	type recipient struct {
+		TransportType string `json:"transportType"`
+		Channel       string `json:"channel"`
+		AblyKey       string `json:"ablyKey"`
+		AblyUrl       string `json:"ablyUrl"`
+	}
+	type push struct {
+		Recipient recipient `json:"recipient"`
+	}
+	type input struct {
+		Id         string `json:"id"`
+		Platform   string `json:"platform"`
+		FormFactor string `json:"formFactor"`
+		Push       push   `json:"push"`
+	}
+	regInput := &input{
+		Id:         deviceID,
+		Platform:   "browser",
+		FormFactor: "other",
+		Push: push{
+			Recipient: recipient{
+				TransportType: "ablyChannel",
+				Channel:       outputChannel,
+				AblyKey:       config.Ably.APIKey,
+				AblyUrl:       config.Subscriber.PushDevice.URL,
+			},
+		},
+	}
+	resp, err := rest.Request(
+		"POST",
+		"/push/deviceRegistrations",
+		nil,
+		regInput,
+		nil)
+	if err != nil {
+		log.Debug("error registering a push device", "err", err)
+		return err
+	}
+	if !resp.Success {
+		log.Debug("error registering a push device", "statusCode", resp.StatusCode, "errorMessage", resp.ErrorMessage)
+		return errors.New(resp.ErrorMessage)
+	}
+	return nil
+}
+
+func subscribePushDevice(config *config.Config, deviceID, channel string, rest *ably.REST, log log15.Logger) error {
+	type input struct {
+		Channel  string `json:"channel"`
+		DeviceId string `json:"deviceId"`
+	}
+	subInput := &input{
+		Channel:  channel,
+		DeviceId: deviceID,
+	}
+	resp, err := rest.Request(
+		"POST",
+		"/push/channelSubscriptions",
+		nil,
+		subInput,
+		nil)
+	if err != nil {
+		log.Debug("error subscribing to a channel", "err", err)
+		return err
+	}
+	if !resp.Success {
+		log.Debug("error subscribing to a channel", "statusCode", resp.StatusCode, "errorMessage", resp.ErrorMessage)
+		return errors.New(resp.ErrorMessage)
+	}
+	return nil
+}
+
 // runSubscriber runs a subscriber task which renders the channel names using
 // the given user number and subscribes to each of them.
 func (l *loadTest) runSubscriber(ctx context.Context, client Client, userNum int64) error {
 	channels := renderChannels(l.subscriberChannels, userNum)
 
+	if l.w.Conf().Subscriber.PushDevice.Enabled {
+		l.log.Debug("creating push device")
+
+		rest, err := ably.NewREST(l.w.conf.Ably.ClientOptions()...)
+		if err != nil {
+			l.log.Debug("error creating a REST client", "err", err)
+			l.w.boomer.RecordFailure("ablyboomer", "createREST", 0, err.Error())
+			return err
+		}
+
+		name := randomString(8)
+		deviceID := fmt.Sprintf("device-%v", name)
+		outputChannel := fmt.Sprintf("push-%v", name)
+
+		err = registerPushDevice(l.w.Conf(), deviceID, outputChannel, rest, l.log)
+		if err != nil {
+			l.w.boomer.RecordFailure("ablyboomer", "registerPushDevice", 0, err.Error())
+		}
```

Comment:
If this fails, then subscribing the device is also going to fail.

We should wrap this in a retry loop with a sleep like we do when trying to subscribe, since as we scale up a load test these registrations may intermittently fail.

Reply:
:+1: , but shouldn't we do the same with the device subscription then?

Reply:
Just add that, as well as an optional push metachannel subscription option. Very useful when debugging push.

Reply:
Yes, sorry, I meant wrap the whole push registration + subscription in a retry loop, which I see you have done 👍